Switch routes for blue green deployments

cchaudier opened the pull request:
## Purpose

Currently, our deployments create new complete stacks including routes to access each stack.
In a functional architecture, we need to manage routes differently to allow uninterrupted service when making a new deployment.

## Proposal

Stop creating routes together with the stack of svc/dc/ep.

Create only 3 routes for each service. The following hosts are examples for the lms:

- previous: "**previous**.lms.{{ project_name }}.{{ domain_name }}"
- current: "lms.{{ project_name }}.{{ domain_name }}"
- next: "**next**.lms.{{ project_name }}.{{ domain_name }}"

Each time we make a new deployment, the following actions are taken:

1) create the new stack,
2) create/update the next route:
    - if the "next" route exists, update it to point to the new stack. 
              - making sure that it is not pointed by the "current" route (it is normally not possible but could happen if something went wrong in a route switch... so let's check that explicitly before deleting it),
              - Delete the stack that was previously pointed by the "next" route, 

    - if the "next" route does not exist, create it and point it to the new stack.

Each time we confirm a deployment by triggering a route switch, the following actions are taken:

1) point the "current" route to the stack that was pointed by the "next" route,
2) point the "previous" route to the stack that was pointed by the "current" route just before we changed it in 1)
3) delete the stack that was pointed by the "previous" route just before we changed it in 2) 
4) patch the next route to the init stamp

## TODO
~~- [x] Create fixed routes on init project~~ now on PR #66 
- [x] Update the next route after deployment
- [x] Delete the previous stack
- [x] Switch routes next to current
- [x] Switch routes current to previous

sampaccoud commented on delete_project.yml:
```diff
@@ -0,0 +1,23 @@
+---
+# This playbook delete an OpenShift project for a particular
```
s/delete/deletes

jmaupetit commented:
Concerning route patterns, I think we can use only the environment type instead of the project name as the domain name is already specific to a customer. Hence, I propose the following pattern:

```
((previous|next).)?lms.{{ env_type }}.{{ domain_name }}
```

Examples:

* https://previous.lms.development.fun-mooc.fr
* https://richie.staging.fun-mooc.fr
* https://next.cms.production.fun-mooc.fr

WDYT?

jmaupetit commented on group_vars/all/main.yml:
```diff
@@ -80,6 +80,18 @@ richie_database:
   port: "{{ postgresql.port }}"
   name: richie
 
+# Openshift services list on the project
```
`s/on/for/`

cchaudier replied:
Done

jmaupetit commented on group_vars/all/main.yml:
```diff
@@ -105,6 +117,10 @@ openshift_routes:
+  - previous
+  - current
+  - next
 openshift_services:
```
This YAML should not validate since you have duplicated variables. This variable should be named `openshift_svcs` to avoid any confusion with what we call a "service" (_i.e._ `svc + dc`)

cchaudier replied:
Done

jmaupetit commented on init_project.yml:
```diff
@@ -24,3 +24,7 @@
     - import_tasks: tasks/create_route_aliases.yml
+      tags: route
+
+    - import_tasks: tasks/create_fixed_routes.yml
```
What about static routes instead of fixed routes?

cchaudier replied:
Yes

jmaupetit commented on tasks/create_fixed_routes.yml:
```diff
@@ -0,0 +1,9 @@
+# Create fixed routes for all apps
+
+- include_tasks: create_fixed_route.yml
+  loop: "{{ openshift_apps }}"
```
:thinking: I think you wanted to name `openshift_services` as `openshift_apps` in your main groupvars, right?

cchaudier replied:
You're right il a copy/paste error :$

sampaccoud commented on group_vars/all/main.yml:
```diff
@@ -105,6 +117,10 @@ openshift_routes:
+openshift_routes_prefixes:
+  - previous
+  - current
+  - next
```
I'm skeptical that this should be configurable. Keep it simple and impose it?

cchaudier replied:
Ok.

sampaccoud commented on tasks/create_fixed_route.yml:
```diff
@@ -0,0 +1,15 @@
+---
+# Create fixed routes only for one apps
```
s/apps/app

cchaudier replied:
done

sampaccoud commented on tasks/create_fixed_route.yml:
```diff
@@ -0,0 +1,15 @@
+  debug: msg="app {{ app | to_yaml }}"
+  tags: route
+
+- name: Make sure the fixed route exist
```
s/exist/exists

cchaudier replied:
done

sampaccoud commented on tasks/create_static_apps_routes.yml:
```diff
@@ -0,0 +1,9 @@
+---
+# Create fixed routes for all apps
```
s/fixed/static

Commit:
🔊(Ansible) use more readable logs

Commit:
✨(route:deploy) update the next route after deploy

if the "next" route exists, update it to point to the new stack.
Delete the stack that was previously pointed by the "next" route.

Commit:
✨(app:deploy) delete the previous stack after deploy

Commit:
✨(route:switch) move current route to the next stack

We confirm a deployment by triggering a route switch:
1. point the previous route to the current stack
2. point the current route to the next stack
3. point the next route to the default stack (init) who don't exist
4. delete the previous stack

jmaupetit commented on docs/developer_guide/playbooks.md:
```diff
@@ -93,6 +93,8 @@ The `deploy.yml` playbook defines a new `deployment_stamp` to create a whole new
 stack by defining OpenShift objects with unique labels and names. This allows us
 to initiate a blue/green deployment strategy.
 
+https://framindmap.org/c/maps/507624/public
```
I'm not sure this diagram is useful as is.

cchaudier replied:
OK.

jmaupetit commented on docs/developer_guide/playbooks.md:
```diff
@@ -106,6 +108,27 @@ $ docker run --rm -it \
+## `switch.yml`
+
+The `switch.yml` playbook move the current stack to the previous route
+and move the next stack to the current route.
```
What about the following:

> The `switch.yml` playbook moves: 1. the "current" stack to the "previous" route, and 2. the "next" stack to the "current" route.

cchaudier replied:
done

jmaupetit commented on deploy.yml:
```diff
@@ -12,24 +12,36 @@
+    # Patch all next routes of all apps after deploy
+    - include_tasks: tasks/run_tasks_for_apps.yml
+      vars:
+        prefixe: "next"
```
`s/prefixe/prefix/`

cchaudier replied:
done

jmaupetit commented on get_route.yml:
```diff
@@ -0,0 +1,19 @@
+    # Patch all current routes of all apps after deploy
+    - include_tasks: tasks/run_tasks_for_apps.yml
+      vars:
+        prefixe: "current"
```
`s/prefixe/prefix/`

cchaudier replied:
done

jmaupetit commented on get_route.yml:
```diff
@@ -0,0 +1,19 @@
+        prefixe: "current"
+        tasks:
+          - deploy_get_stamp_from_route
+          - deploy_patch_route
```
We might consider creating a folder per purpose in `tasks`, _i.e._ `tasks/deploy/` instead of prefixing tasks by purpose (_i.e._ `deploy_*`). This would make things more clear/understandable.

cchaudier replied:
I propose to do discuss that on a issue because the relative paths can can reduce the lisibility of the tasks.

jmaupetit commented on plugins/callback/anstomlog.py:
```diff
@@ -0,0 +1,395 @@
+# coding=utf-8
```
Where does this plugin come from? What is the license attached to this code?

cchaudier replied:
it cam from https://github.com/octplane/ansible_stdout_compact_logger with MIT licence.
I've comment the file

jmaupetit commented on switch.yml:
```diff
@@ -0,0 +1,22 @@
+---
+# This playbook switch all routes:
+# 1. point the "current" route to the stack that was pointed by the "next" route,
+# 2. point the "previous" route to the stack that was pointed by the "current" route just before we changed it in 1)
```
For clarity, point 2. must be 1. (and reverse).

jmaupetit replied:
Also you can remove "just before we changed it in 1)"

cchaudier replied:
done

jmaupetit commented on switch.yml:
```diff
@@ -0,0 +1,22 @@
+# This playbook switch all routes:
+# 1. point the "current" route to the stack that was pointed by the "next" route,
+# 2. point the "previous" route to the stack that was pointed by the "current" route just before we changed it in 1)
+# 3. delete the stack that was pointed by the "previous" route just before we changed it in 2)
```
remove "just before we changed it in 2)"

cchaudier replied:
done

jmaupetit commented on switch.yml:
```diff
@@ -0,0 +1,22 @@
+# 1. point the "current" route to the stack that was pointed by the "next" route,
+# 2. point the "previous" route to the stack that was pointed by the "current" route just before we changed it in 1)
+# 3. delete the stack that was pointed by the "previous" route just before we changed it in 2)
+# 4. patch the next route to the init stamp
```
Can we delete the "next" route instead if it's not pointing to an available stack?

cchaudier replied:
For me it's a static route I prefere do not delete the route.

jmaupetit commented on switch.yml:
```diff
@@ -0,0 +1,22 @@
+
+    - import_tasks: tasks/set_vars.yml
+
+    # Patch all apps dest routes with src stack
```
This comment is confusing to me :wink:

cchaudier replied:
great move it's an old comment ;)

jmaupetit commented on tasks/deploy_apps.yml:
```diff
@@ -0,0 +1,17 @@
+---
```
The name of this task set is confusing and does not reveal what it does. What about naming it `manage_apps.yml` or alike?

cchaudier replied:
I going to do this **after** including your PR.

jmaupetit commented on tasks/deploy_apps.yml:
```diff
@@ -0,0 +1,17 @@
+#   deployment_state:
+#     - present : default value, all objects are created
+#     - 'absent': all objects are deleted
+#   deployment_stamp : the stamp of the object we going to creat or delete
```
What about the following?

```yaml
# Task set to manage apps
# 
# Args:
#   deployment_state:
#     - "present": default value, all objects are created
#     - "absent":  all objects are deleted
#   deployment_stamp: the stamp of the object we are going to create or delete
```

FTR: there is no space before a colon (`:`) in English :wink:

cchaudier replied:
ok

jmaupetit commented on tasks/deploy_get_stamp_from_route.yml:
```diff
@@ -0,0 +1,14 @@
+---
+# Get deployment_stamp from a route
+
+- name: Get deployment_stamp of route[{{ app.name }}-{{ prefixe }}]
```
`s/of/from/`

jmaupetit commented on tasks/deploy_get_stamp_from_route.yml:
```diff
@@ -0,0 +1,14 @@
+
+- name: Get deployment_stamp of route[{{ app.name }}-{{ prefixe }}]
+  openshift_raw:
+    definition: "{{ lookup('template', 'templates/openshift/common/utils/get_deployment_stamp_form_route.yml.j2') | from_yaml }}"
```
`[...]_form_route.yml.j2` -> `[...]_from_route.yml.j2`?

cchaudier replied:
OMG !
Done ;)

jmaupetit commented on templates/openshift/common/utils/get_deployment_stamp_form_route.yml.j2:
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: Route
+metadata:
+  name: "{{ app.name }}-{{ prefixe }}"
```
`s/prefixe/prefix`

cchaudier replied:
done

jmaupetit commented on tasks/deploy_get_stamp_from_route.yml:
```diff
@@ -0,0 +1,14 @@
+- name: Get deployment_stamp of route[{{ app.name }}-{{ prefixe }}]
+  openshift_raw:
+    definition: "{{ lookup('template', 'templates/openshift/common/utils/get_deployment_stamp_form_route.yml.j2') | from_yaml }}"
+  register: my_route
```
I am sure we can work on this variable name to improve its meaning. What about `app_route`?

cchaudier replied:
done

jmaupetit commented on tasks/deploy_get_stamp_from_route.yml:
```diff
@@ -0,0 +1,14 @@
+
+- name: set_fact route_deployment_stamp
+  set_fact:
+    route_deployment_stamp: "{{ my_route.result.metadata.labels.deployment_stamp }}"
```
`s/route_deployment_stamp/app_route_deployment_stamp/`?

cchaudier replied:
Done

jmaupetit commented on tasks/deploy_patch_route.yml:
```diff
@@ -0,0 +1,12 @@
+---
+# Patch routes of all services of one app to the new deployment_stamp
```
> Patch routes of all services from a single app to the new deployment_stamp

cchaudier replied:
done

jmaupetit commented on tasks/deploy_patch_route.yml:
```diff
@@ -0,0 +1,12 @@
+    state: present
+  with_nested:
+    - "{{ app.services }}"
+    - "{{ prefixe }}"
```
`s/prefixe/prefix/`

cchaudier replied:
done

jmaupetit commented on tasks/switch_route.yml:
```diff
@@ -0,0 +1,22 @@
+---
+# Patch all apps dest routes with src stack
```
Still confusing comment :smile:

cchaudier replied:
Now the comment is : `# Patch app dest routes with src stack`

jmaupetit commented on tasks/switch_route.yml:
```diff
@@ -0,0 +1,22 @@
+- include_tasks: deploy_patch_route.yml
+  vars:
+    prefixe: "{{ prefixe_route_src }}"
+  tags: switch
```
`s/prefixe_*/prefix_*/g`

cchaudier replied:
done

jmaupetit commented on tasks/switch_routes.yml:
```diff
@@ -0,0 +1,29 @@
+    prefixe: "previous"
+  tags: switch
+
+- name: save previous_route_deployment_stamp
```
`s/save/Save`

cchaudier replied:
done

jmaupetit commented on tasks/switch_route.yml:
```diff
@@ -0,0 +1,22 @@
+- include_tasks: deploy_patch_route.yml
+  vars:
+    prefixe: "{{ prefixe_route_src }}"
+  tags: switch
```
Also, this won't work for the "next" route has there is no stack to point to.

cchaudier replied:
It work because on template the default deployment_stamp is init

jmaupetit replied:
I've removed this the default `deployment_stamp` value is `null`.

jmaupetit commented on templates/openshift/common/route/main.yml.j2:
```diff
@@ -6,7 +6,7 @@ metadata:
-    app:              "{{ app.name }}-{{ deployment_stamp | default('init', true) }}"
+    app:              "{{ app.name }}"
     version:          "{{ item.0.tag | default('init', true) }}"
     deployment_stamp: "{{ deployment_stamp | default('init', true) }}"
```
The `init` deployment stamp does not exist anymore.

cchaudier replied:
Yes I now, I can change for `null` if you want.

jmaupetit reviewed (commented):
Thanks for this amazing work. One question remains: what to do with the next route? What is achieved here is still unclear to me. IMO we need to remove this "next" route since it will be re-created by the next deployment.

Commit:
👌(typo) updating code due to code review changes

Commit:
🔊(logs) improve logs

Commit:
🏗(vars) make tempory app vars private

The app variable name is use on to different task. In set_vars we make it private to prevent tis warning :
system | included: /app/tasks/run_tasks_for_apps.yml for local
include_tasks ... [WARNING]: The loop variable 'app' is already in use. You should set the `loop_var` value in the `loop_control` option for the task to something else to avoid variable collisions and unexpected behavior.

Commit:
🔥(routes) static routes are managed by a _common template

We don't need this templates

Commit:
♻️(deploy) rename task deploy_apps to manage_app

Now the task file create or destroy only one app so manage_app is a better name

Commit:
♻️(hello) move hello on availble apps

i need use hello to develop because it's a very light app.

Commit:
🔖(arnold) upgrade version number

jmaupetit commented on tasks/manage_app.yml:
```diff
@@ -20,37 +26,33 @@
   tags:
     - deploy
-    - deployment
+    - deployments
```
What is the purpose of having `deploy` and `deployments` tags?

cchaudier replied:
Sorry it's your code : https://github.com/openfun/arnold/blob/master/tasks/deploy_app.yml
But If whe want we can filter the deploy only on deployment or service or whatever.

jmaupetit replied:
We need to make a clean up PR to make our tags consistent over tasks.

cchaudier replied:
I'm agree :+1:

jmaupetit commented on tasks/manage_app.yml:
```diff
@@ -1,12 +1,18 @@
 ---
-# Deploy an app
+# Task set to manage apps
```
Can we call this file `manage_app_services.yml` instead?

cchaudier replied:
For me on app have a lot of services.
All over tasks who interact with app and services end with app on their names.

jmaupetit replied:
I don't get your point: I want to move the `tasks/manage_app.yml` to `tasks/manage_app_services.yml` because "manage app" is too generic. Maybe "manage app services" is not relevant, but I have no better idea at the time. What about "manage app objects"?

cchaudier replied:
By definition one app is a stak : a collection of services and ressources.
This playbook manage one app, for me the name is good if we have a good doc.
Why do not rename `delete_app.yml` by `delete_app_objects.yml` if we rename this manage playbook ?

jmaupetit replied:
Okay let's do this!

jmaupetit commented:
:clap: great work btw
